<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Tutorial: Basic statistics in R</title>

<script src="Basic_statistics_tutorial_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="Basic_statistics_tutorial_files/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="Basic_statistics_tutorial_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="Basic_statistics_tutorial_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="Basic_statistics_tutorial_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="Basic_statistics_tutorial_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="Basic_statistics_tutorial_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="Basic_statistics_tutorial_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="Basic_statistics_tutorial_files/navigation-1.1/tabsets.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { background-color: #f8f8f8; }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Tutorial: Basic statistics in R</h1>

</div>


<!-- rmd lab header -->
<!-- custom fonts -->
<p><link href="https://fonts.googleapis.com/css?family=Roboto|Source+Sans+Pro:300,400,600|Ubuntu+Mono&amp;subset=latin-ext" rel="stylesheet"></p>
<p><br></p>
<div id="a-quick-catch-up-and-practice-for-previous-sessions" class="section level1">
<h1><span class="header-section-number">1</span> A quick catch up and practice for previous sessions</h1>
<div id="data-import-export-generation-vectors-matrices-and-data-frames" class="section level2">
<h2><span class="header-section-number">1.1</span> Data: Import, Export, Generation, vectors, matrices and data frames</h2>
<p>For this session we will use Gene counts data from Lokesh given in the previous session. You can download the data here <a href="https://immuntech.github.io/Bioinformatics_Workshop_ImmTech/Data/2019-10-15/gene_counts.tsv">genes</a> and <a href="https://immuntech.github.io/Bioinformatics_Workshop_ImmTech/Data/2019-10-15/metadata.tsv">metadata</a>.</p>
Please load the data into R.
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="co"># Please type the correct path to where you save your data files.</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2"></a>
<a class="sourceLine" id="cb1-3" data-line-number="3">gen_counts&lt;-<span class="kw">read.table</span>(<span class="st">&quot;~/Documents/GitHub/Bioinformatics_Workshop_ImmTech/Data/2019-11-18/data/gene_counts.tsv&quot;</span>)</a>
<a class="sourceLine" id="cb1-4" data-line-number="4">gen_counts &lt;-<span class="kw">as.matrix</span>(gen_counts)</a>
<a class="sourceLine" id="cb1-5" data-line-number="5">pheno &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;~/Documents/GitHub/Bioinformatics_Workshop_ImmTech/Data/2019-11-18/data/metadata.tsv&quot;</span>, <span class="dt">header =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
</div>
<p>Have a look at the data by using <code>head()</code> function.</p>
<p>First have look at the the gene expression data.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co"># To look at the top three lines of the file use head()</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="kw">head</span>(gen_counts, <span class="dv">3</span>)</a></code></pre></div>
<pre><code>##         Sample_1 Sample_2 Sample_3 Sample_4 Sample_5 Sample_6 Sample_7
## TMEM240       84       85        0        0        0        0        0
## MT-CO1         0       67        0        0       80      165        0
## ADGRG6         0        0        0        0        0        0       24
##         Sample_8 Sample_9 Sample_10 Sample_11 Sample_12 Sample_13
## TMEM240        0       46         0        38         0         0
## MT-CO1        70      209       130         0       113         0
## ADGRG6         0        0         0         0         0         0
##         Sample_14 Sample_15 Sample_16
## TMEM240         0        72         0
## MT-CO1        276         0         0
## ADGRG6          9         0         0</code></pre>
</div>
<p>And now the Phenotype data.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="kw">head</span>(pheno)</a></code></pre></div>
<pre><code>##     Sample Healthy Age Sex
## 1 Sample_1     yes  23   F
## 2 Sample_2     yes  47   M
## 3 Sample_3     yes  35   F
## 4 Sample_4     yes  19   M
## 5 Sample_5     yes  36   F
## 6 Sample_6     yes  29   M</code></pre>
</div>
<p>It is good practice to keep the annotation is separate data structures. It is often the the one needs to make use of different sample annotations and row annotations for statistics and plotting purpose. Usually the numeric data such as gene expression values is kept in a matrix and all the annotations such as gene annotations and sample annotations are kept in data frames.We might want to recall that Matrix stores only one type of data such as numeric, strings etc. while data frames can hold different types of values.</p>
<p><strong>On your own</strong></p>
<p><em>Try the <code>str()</code>, <code>summary()</code>, <code>dim()</code> and <code>class</code> functions and see if you can find out how many rows and columns each data has. Also find what other attributes you can understand using these functions</em></p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw">summary</span>(pheno)</a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="kw">str</span>(gen_counts)</a>
<a class="sourceLine" id="cb6-3" data-line-number="3"><span class="kw">class</span>(gen_counts)</a>
<a class="sourceLine" id="cb6-4" data-line-number="4"><span class="kw">class</span>(pheno)</a></code></pre></div>
</div>
<div id="accessing-parts-of-the-data" class="section level3">
<h3><span class="header-section-number">1.1.1</span> Accessing parts of the data</h3>
<p>R is a flexible language and it gives many different options to do the same thing. Something simple like accession certain rows of the data or columns of the can be done in many different ways. Lets look do it by using <em>positional indices mapping</em> or
by using row or column <em>names</em>.</p>
<p><strong>Using positional indices</strong></p>
<p>Accessing parts of the data in R can be done in different ways. Lets start by using <code>data[ , ]</code> to extract parts of the data. Use left side of the comma for rows and right side of the comma for columns.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">gen_counts[ <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, ]</a>
<a class="sourceLine" id="cb7-2" data-line-number="2">gen_counts[ <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</a>
<a class="sourceLine" id="cb7-3" data-line-number="3">gen_counts[ , <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</a></code></pre></div>
</div>
<p>Note that if you leave either side of the comma as blank, r assumes that you meant all data in that dimension should be selected.</p>
<p><strong>Using names</strong></p>
<p>Lets say you are only interested in certain samples and you want to extract those by names.
Lets create a vector of samples you are interested in. To extract the samples from your data simply write the name of the vector on the right side of the comma in the square brackets <code>[ , ]</code>.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="co">## Create a vector of samples you are interested in</span></a>
<a class="sourceLine" id="cb8-2" data-line-number="2">samples &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Sample_13&quot;</span> , <span class="st">&quot;Sample_9&quot;</span> , <span class="st">&quot;Sample_4&quot;</span> , <span class="st">&quot;Sample_14&quot;</span> , <span class="st">&quot;Sample_12&quot;</span>)</a>
<a class="sourceLine" id="cb8-3" data-line-number="3"></a>
<a class="sourceLine" id="cb8-4" data-line-number="4"><span class="co">## To extract the samples, here I am using indices for rows while names for columns.</span></a>
<a class="sourceLine" id="cb8-5" data-line-number="5"></a>
<a class="sourceLine" id="cb8-6" data-line-number="6">gen_counts[ <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, samples]</a></code></pre></div>
<pre><code>##         Sample_13 Sample_9 Sample_4 Sample_14 Sample_12
## TMEM240         0       46        0         0         0
## MT-CO1          0      209        0       276       113
## ADGRG6          0        0        0         9         0</code></pre>
</div>
<p>And similarly, to extract you favorite genes by names use following examples.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">genes &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;BMS1&quot;</span> , <span class="st">&quot;SYNC&quot;</span> , <span class="st">&quot;NFIB&quot;</span> , <span class="st">&quot;RAB11FIP3&quot;</span> , <span class="st">&quot;COG1&quot;</span> , <span class="st">&quot;C5orf38&quot;</span> , <span class="st">&quot;FAM126B&quot;</span> ,   <span class="st">&quot;SUCLA2&quot;</span> , <span class="st">&quot;SCN4A&quot;</span> , <span class="st">&quot;CISD1&quot;</span>  )</a>
<a class="sourceLine" id="cb10-2" data-line-number="2"></a>
<a class="sourceLine" id="cb10-3" data-line-number="3"><span class="co">## To extract the rows, here I am using indices for rows while names for columns.</span></a>
<a class="sourceLine" id="cb10-4" data-line-number="4"></a>
<a class="sourceLine" id="cb10-5" data-line-number="5"><span class="kw">head</span>(gen_counts[genes,], <span class="dv">3</span>)</a></code></pre></div>
<pre><code>##      Sample_1 Sample_2 Sample_3 Sample_4 Sample_5 Sample_6 Sample_7
## BMS1      146        0      253        0        0      152      119
## SYNC       63       77      120       47       43       25       11
## NFIB       34        0        0       56        0        0        0
##      Sample_8 Sample_9 Sample_10 Sample_11 Sample_12 Sample_13 Sample_14
## BMS1        0       49       198         0        73         0       134
## SYNC       35       38        24       130        26         0         0
## NFIB        0        0         0         0         0         0         0
##      Sample_15 Sample_16
## BMS1       342         0
## SYNC        37         0
## NFIB         0         0</code></pre>
</div>
<p>**Extract names samples coming from healthy and ill individuals.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1">healthy &lt;-<span class="st"> </span>pheno[pheno<span class="op">$</span>Healthy<span class="op">==</span><span class="st">&quot;yes&quot;</span>, ]<span class="op">$</span>Sample</a>
<a class="sourceLine" id="cb12-2" data-line-number="2">ill &lt;-<span class="st"> </span>pheno[pheno<span class="op">$</span>Healthy<span class="op">==</span><span class="st">&quot;no&quot;</span>, ]<span class="op">$</span>Sample</a></code></pre></div>
</div>
</div>
<div id="save-the-extracted-data" class="section level3">
<h3><span class="header-section-number">1.1.2</span> Save the extracted data</h3>
<p>So far we have only printer the extracted data in the console, this is not very useful if we want to use the extracted data for further analyses. Let us store this data in the “environment” by using <code>&lt;-</code> or <code>=</code> signs and write it to a file with <code>write.table()</code> function.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">gen_counts_part &lt;-<span class="st"> </span>gen_counts[genes, ] </a>
<a class="sourceLine" id="cb13-2" data-line-number="2"></a>
<a class="sourceLine" id="cb13-3" data-line-number="3"><span class="kw">write.table</span>(<span class="dt">x =</span> gen_counts_part, <span class="dt">file =</span> <span class="st">&quot;~/Documents/GitHub/Bioinformatics_Workshop_ImmTech/Data/2019-11-18/data/part_gene_counts.tsv&quot;</span>, <span class="dt">row.names =</span> <span class="ot">TRUE</span>, <span class="dt">col.names =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
</div>
<p>So far we have consolidated our knowledge on how to load, access, subset and save data. This is going to be useful in later session.</p>
</div>
</div>
</div>
<div id="functional-programming" class="section level1">
<h1><span class="header-section-number">2</span> Functional programming</h1>
<p>R is a <strong>Functional Language</strong> and there is a lot of help available. For almost all necessary data analyses tasks, there are functions available. Most of these functions are well written and have good documentation on how to use them. Often R functions are bundled in packages. An example of a package is <code>ggplot2</code> that is developed for plotting purpose. There are plenty more such as <code>limma</code> for analyses of micro array data. There are handy functions to find out what these packages and functions do.
We would take a look at how to get help on packages, functions and data types.</p>
<div id="help-on-functions" class="section level3">
<h3><span class="header-section-number">2.0.1</span> Help on functions</h3>
<p><strong><em>And to get help on <code>ggplot()</code> function lets use <code>?</code>. Or <code>t.test()</code> or <code>lm()</code>.</em></strong></p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1">?ggplot</a>
<a class="sourceLine" id="cb14-2" data-line-number="2">?t.test</a>
<a class="sourceLine" id="cb14-3" data-line-number="3">?lm</a></code></pre></div>
</div>
</div>
<div id="help-on-r-packages" class="section level3">
<h3><span class="header-section-number">2.0.2</span> Help on R packages</h3>
<p><strong><em>To know more about packages, lets look at the Vignettes of ggplot2 package in R by using function call <code>browseVignettes(&quot;ggplot2&quot;)</code>.</em></strong></p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1"><span class="kw">browseVignettes</span>(<span class="st">&quot;ggplot2&quot;</span>)</a></code></pre></div>
</div>
</div>
<div id="help-on-data-closeness." class="section level3">
<h3><span class="header-section-number">2.0.3</span> Help on data closeness.</h3>
<p><strong><em>If you want to know what methods are available for your type of data, you can use <code>methods</code> function as following.</em></strong></p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1"><span class="co"># For matrix class</span></a>
<a class="sourceLine" id="cb16-2" data-line-number="2"><span class="kw">methods</span>(<span class="dt">class =</span> <span class="st">&quot;matrix&quot;</span>)</a>
<a class="sourceLine" id="cb16-3" data-line-number="3"></a>
<a class="sourceLine" id="cb16-4" data-line-number="4"><span class="co"># For data frames ()</span></a>
<a class="sourceLine" id="cb16-5" data-line-number="5"><span class="kw">methods</span>(<span class="dt">class =</span> <span class="st">&quot;data.frame&quot;</span>)</a></code></pre></div>
</div>
<p>Essentially each function as a set of parameters you can give. Often the errors R produces is due to the fact that the user is not providing the correct data (arguments) for those parameters. Help pages will give an overview of the parameters available. Typically there are default values (arguments) for each parameter should be but the function wont work with some minimal arguments.</p>
</div>
</div>
<div id="descriptive-statistics" class="section level1">
<h1><span class="header-section-number">3</span> Descriptive statistics</h1>
<p>R is a statistical and programming language.There are thousands of packages and functions available for basic to very advanced statistics. In not, once can always write there own function to help themselves in their cause.</p>
<div id="some-useful-descriptive-statistics" class="section level2">
<h2><span class="header-section-number">3.1</span> Some useful descriptive statistics</h2>
<p>There are several convenience functions and packages available for descriptive statistics and analyses in R. We will use some of these and try to apply them on our genes data.</p>
<p>To get values such as mean, standard deviation, variance, min, max, median etc. R has implemented functions in the base package. This package is automatically loaded in R when you start the program R or Studio.</p>
<p>Lets say we want obtain the mean of the fist sample we can use <code>rowMeans()</code> , <code>colMeans()</code>, <code>aggregate()</code> as following.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="co"># means on selection of samples</span></a>
<a class="sourceLine" id="cb17-2" data-line-number="2"><span class="kw">colMeans</span>(gen_counts[ , samples])</a>
<a class="sourceLine" id="cb17-3" data-line-number="3"></a>
<a class="sourceLine" id="cb17-4" data-line-number="4"><span class="co"># means of selection of rows and columns. </span></a>
<a class="sourceLine" id="cb17-5" data-line-number="5"></a>
<a class="sourceLine" id="cb17-6" data-line-number="6"><span class="kw">rowMeans</span>(gen_counts[genes , samples])</a></code></pre></div>
<pre><code>## Sample_13  Sample_9  Sample_4 Sample_14 Sample_12 
##  43.28804  56.36232   9.75000  45.61957  23.19203 
##      BMS1      SYNC      NFIB RAB11FIP3      COG1   C5orf38   FAM126B 
##      51.2      22.2      11.2      14.0       3.0      41.6      75.0 
##    SUCLA2     SCN4A     CISD1 
##      53.0      53.4       0.0</code></pre>
</div>
<p>Similarly, we can apply the same other descriptive statistical functions such as <code>rowVars()</code> or <code>colVars()</code> for variances of rows and columns respectively.</p>
<p><strong>Try to use <code>rowsd, rowMax, rowMin</code> etc. to get an idea of the functions available for descriptive statistics.</strong></p>
<p>If there is a particular function you are interested in, ask one of the TAs or brown R related webpages.</p>
<p><strong><em>Mean, Median, Variance, Standard Deviation, Quantiles, counts, log2, log10, Exponential</em></strong></p>
</div>
<div id="applying-functions-on-subset-of-the-data" class="section level2">
<h2><span class="header-section-number">3.2</span> Applying functions on subset of the data</h2>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1"><span class="kw">aggregate</span>(Age<span class="op">~</span>Sex,<span class="dt">data =</span> pheno, <span class="dt">FUN =</span> <span class="st">&quot;mean&quot;</span>)</a>
<a class="sourceLine" id="cb19-2" data-line-number="2"><span class="kw">aggregate</span>(Age<span class="op">~</span>Sex,<span class="dt">data =</span> pheno, <span class="dt">FUN =</span> <span class="st">&quot;sd&quot;</span>)</a></code></pre></div>
<pre><code>##   Sex    Age
## 1   F 52.625
## 2   M 43.500
##   Sex      Age
## 1   F 24.43614
## 2   M 22.27106</code></pre>
</div>
</div>
</div>
<div id="writing-your-own-functions-apply-family-functions" class="section level1">
<h1><span class="header-section-number">4</span> Writing your own functions <code>apply()</code>*** family functions</h1>
<p>Writing function is relatively easy. Lets write a function to calculate ratio of means between healthy vs. ill individuals.</p>
<div id="writing-your-own-functions" class="section level2">
<h2><span class="header-section-number">4.1</span> Writing your own functions</h2>
<p>We define <code>ratioOfMeans</code> by assigning it to the output of <em>function</em>. The list of <em>argument</em> names are contained within parentheses. Next, the <em>body</em> of the function–the statements that are executed when it runs–is contained within curly braces ({}).</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1"><span class="co"># This functions has three parameters, names of group1 and names of group 2.</span></a>
<a class="sourceLine" id="cb21-2" data-line-number="2">ratioOfMeans &lt;-<span class="st"> </span><span class="cf">function</span>(data, group1, group2) {</a>
<a class="sourceLine" id="cb21-3" data-line-number="3">  mean1 &lt;-<span class="st"> </span><span class="kw">rowMeans</span>(data[ , group1])</a>
<a class="sourceLine" id="cb21-4" data-line-number="4">  mean2 &lt;-<span class="st"> </span><span class="kw">rowMeans</span>(data[ , group2])</a>
<a class="sourceLine" id="cb21-5" data-line-number="5">  ratio &lt;-<span class="st"> </span>mean1<span class="op">/</span>mean2</a>
<a class="sourceLine" id="cb21-6" data-line-number="6">  <span class="kw">return</span>(ratio)</a>
<a class="sourceLine" id="cb21-7" data-line-number="7">}</a></code></pre></div>
</div>
<p>When we call the function, the values we pass to it are assigned to those variables so that we can use them inside the function. Inside the function, we use a return statement to send a result back to whoever asked for it.</p>
<p>Now we have written a function, we can use it to calculate ratios within our own data sets whenever we want.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1">ratios &lt;-<span class="st"> </span><span class="kw">ratioOfMeans</span>(<span class="dt">data =</span> gen_counts, <span class="dt">group1 =</span> healthy, <span class="dt">group2 =</span> ill)</a></code></pre></div>
</div>
<p>*** It is OK if you don’t fully understand this, for typical data analyses one can usually rely on already available functions in R***. There is a lot of testing and annotation required before your function can be published for others to use.
This was just to illustrate the value of investing time in learning R.</p>
</div>
<div id="the-apply-family-functions" class="section level2">
<h2><span class="header-section-number">4.2</span> The <strong><em><code>apply()</code></em></strong> family functions</h2>
<p>R has a special class of helper functions that can be used to apply functions to all elements of any data type. These include <code>lapply()</code>, <code>sapply()</code>, <code>maapply()</code>, <code>tapply()</code>, <code>rapply()</code>. These are alternatives to loops and have faster implementation. It is recommended to use these instead of loops as they are much faster and rather easy to implement.</p>
<p>Lets take a look at an example.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1"><span class="co"># To calculate mean of rows use following</span></a>
<a class="sourceLine" id="cb23-2" data-line-number="2"><span class="kw">apply</span>(gen_counts, <span class="dv">1</span>, mean)</a>
<a class="sourceLine" id="cb23-3" data-line-number="3"><span class="co"># To calculate mean of columns use following</span></a>
<a class="sourceLine" id="cb23-4" data-line-number="4"><span class="kw">apply</span>(gen_counts, <span class="dv">2</span>, mean)</a></code></pre></div>
</div>
<p>The above function should give you the exact same output as <code>rowMeans()</code> and <code>colMeans()</code> respectively.</p>
<p>The <code>apply()</code> functions are particularly handy when you want to write small <strong><em>through away functions</em></strong>.</p>
<p>We will use these in the upcoming session when we want to learn about t.test and linear models.</p>
<p><strong><em>We will cover more advanced descriptive statistics in the future sessions.</em></strong></p>
</div>
</div>
<div id="basic-statistics" class="section level1">
<h1><span class="header-section-number">5</span> Basic Statistics</h1>
<p>We will look at the two different statistical tests namely <em>t-test</em> and <em>linear regression</em> to test our hypotheses.</p>
<div id="student-t-test" class="section level2">
<h2><span class="header-section-number">5.1</span> Student t-test</h2>
<p>To apply <code>t.test()</code> in R we can first look at the help by typing <code>?t.test</code>.
We want to find out whether certain genes are deferentially expressed in healthy vs. ill patients using students t-test.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1"><span class="co">## Extract the names of the ill samples</span></a>
<a class="sourceLine" id="cb24-2" data-line-number="2">healthy &lt;-<span class="st"> </span>pheno[pheno<span class="op">$</span>Healthy <span class="op">==</span><span class="st"> &quot;yes&quot;</span> , ]<span class="op">$</span>Sample</a>
<a class="sourceLine" id="cb24-3" data-line-number="3">ill &lt;-<span class="st"> </span>pheno[pheno<span class="op">$</span>Healthy <span class="op">==</span><span class="st"> &quot;no&quot;</span>, ]<span class="op">$</span>Sample</a>
<a class="sourceLine" id="cb24-4" data-line-number="4"></a>
<a class="sourceLine" id="cb24-5" data-line-number="5"><span class="co"># Aplly t.test on gene at row 1 in the data.</span></a>
<a class="sourceLine" id="cb24-6" data-line-number="6">t_res&lt;-<span class="st"> </span><span class="kw">t.test</span>(gen_counts[<span class="dv">1</span> , healthy], gen_counts[<span class="dv">1</span> , ill])</a>
<a class="sourceLine" id="cb24-7" data-line-number="7"></a>
<a class="sourceLine" id="cb24-8" data-line-number="8"><span class="co"># Take a look at the resuls of the t test.</span></a>
<a class="sourceLine" id="cb24-9" data-line-number="9">t_res</a></code></pre></div>
<pre><code>## 
##  Welch Two Sample t-test
## 
## data:  gen_counts[1, healthy] and gen_counts[1, ill]
## t = 1.1869, df = 13.681, p-value = 0.2555
## alternative hypothesis: true difference in means is not equal to 0
## 95 percent confidence interval:
##  -15.71249  54.46249
## sample estimates:
## mean of x mean of y 
##    30.000    10.625</code></pre>
</div>
<p>Lets have a look at the output of the t_res object. First have a look at the object by using <code>str()</code> function.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" data-line-number="1"><span class="kw">str</span>(t_res)</a></code></pre></div>
<pre><code>## List of 10
##  $ statistic  : Named num 1.19
##   ..- attr(*, &quot;names&quot;)= chr &quot;t&quot;
##  $ parameter  : Named num 13.7
##   ..- attr(*, &quot;names&quot;)= chr &quot;df&quot;
##  $ p.value    : num 0.255
##  $ conf.int   : num [1:2] -15.7 54.5
##   ..- attr(*, &quot;conf.level&quot;)= num 0.95
##  $ estimate   : Named num [1:2] 30 10.6
##   ..- attr(*, &quot;names&quot;)= chr [1:2] &quot;mean of x&quot; &quot;mean of y&quot;
##  $ null.value : Named num 0
##   ..- attr(*, &quot;names&quot;)= chr &quot;difference in means&quot;
##  $ stderr     : num 16.3
##  $ alternative: chr &quot;two.sided&quot;
##  $ method     : chr &quot;Welch Two Sample t-test&quot;
##  $ data.name  : chr &quot;gen_counts[1, healthy] and gen_counts[1, ill]&quot;
##  - attr(*, &quot;class&quot;)= chr &quot;htest&quot;</code></pre>
</div>
<p>We can extract desired output such as p value, means of groups etc. from this object by using <code>$</code> operator.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" data-line-number="1">t_res<span class="op">$</span>p.value</a></code></pre></div>
<pre><code>## [1] 0.255458</code></pre>
</div>
<p>Try to extract confidence intervals, means of groups and from the same object.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" data-line-number="1">t_res<span class="op">$</span>estimate[<span class="dv">1</span>]</a>
<a class="sourceLine" id="cb30-2" data-line-number="2">t_res<span class="op">$</span>estimate[<span class="dv">2</span>]</a>
<a class="sourceLine" id="cb30-3" data-line-number="3">t_res<span class="op">$</span>conf.int</a></code></pre></div>
<pre><code>## mean of x 
##        30 
## mean of y 
##    10.625 
## [1] -15.71249  54.46249
## attr(,&quot;conf.level&quot;)
## [1] 0.95</code></pre>
</div>
</div>
<div id="many-t-tests-in-one-go." class="section level2">
<h2><span class="header-section-number">5.2</span> Many t tests in one go.</h2>
<p>We can make use of R’s apply function to perform t test on all genes in the data.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" data-line-number="1"><span class="kw">apply</span>(gen_counts , <span class="dv">1</span> , <span class="cf">function</span>(x) <span class="kw">t.test</span>(x[healthy], x[ill]))</a></code></pre></div>
</div>
<p>For nicer output lets try to access p-values.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb33-1" data-line-number="1"><span class="kw">apply</span>(gen_counts,<span class="dv">1</span>,<span class="cf">function</span>(x) <span class="kw">t.test</span>(x[healthy], x[ill])<span class="op">$</span>p.value)</a></code></pre></div>
</div>
<p>And means by using following code.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb34-1" data-line-number="1"><span class="kw">apply</span>(gen_counts,<span class="dv">1</span>,<span class="cf">function</span>(x) <span class="kw">t.test</span>(x[healthy], x[ill])<span class="op">$</span>estimate[<span class="dv">1</span>])</a>
<a class="sourceLine" id="cb34-2" data-line-number="2"><span class="kw">apply</span>(gen_counts,<span class="dv">1</span>,<span class="cf">function</span>(x) <span class="kw">t.test</span>(x[healthy], x[ill])<span class="op">$</span>estimate[<span class="dv">2</span>])</a></code></pre></div>
</div>
<p><strong>Can you extract confidence intervals?</strong></p>
</div>
<div id="linear-regression" class="section level2">
<h2><span class="header-section-number">5.3</span> Linear regression</h2>
<p>Similar to t.test, one can also apply linear regression on a single gene. Take a look at some examples on how to extract output from the models. use <code>?lm</code> and <code>methods(class=&quot;lm&quot;)</code> to know more about what you can extract from linear model fit.</p>
<p>One should be aware the <code>formula</code> argument of the function and it works such that the outcome variable is on the left hand side of the “<code>~</code>” sign and the predictor variables are on the right hand side.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb35-1" data-line-number="1"> <span class="co"># Outcome</span></a>
<a class="sourceLine" id="cb35-2" data-line-number="2">y&lt;-pheno<span class="op">$</span>Age</a>
<a class="sourceLine" id="cb35-3" data-line-number="3"></a>
<a class="sourceLine" id="cb35-4" data-line-number="4"><span class="co"># Model</span></a>
<a class="sourceLine" id="cb35-5" data-line-number="5">lm1&lt;-<span class="st"> </span><span class="kw">lm</span>( <span class="dt">formula =</span> y <span class="op">~</span><span class="st"> </span>gen_counts[<span class="dv">1</span>,])</a></code></pre></div>
</div>
<p>To look at the results from the linear model <code>summary()</code> is a very useful function and it gives most of the output needed for interpretation of results.</p>
<p>To find out what other functions you can apply to linear model output use two of our favorite functions <code>class()</code> and <code>methods()</code>.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" data-line-number="1"><span class="co"># to find out which class of object lm1 is use following</span></a>
<a class="sourceLine" id="cb36-2" data-line-number="2"><span class="kw">class</span>(lm1)</a></code></pre></div>
<pre><code>## [1] &quot;lm&quot;</code></pre>
</div>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb38-1" data-line-number="1"><span class="co"># To find out which methods are availabe for this class</span></a>
<a class="sourceLine" id="cb38-2" data-line-number="2"></a>
<a class="sourceLine" id="cb38-3" data-line-number="3"><span class="kw">methods</span>(<span class="dt">class =</span> <span class="st">&quot;lm&quot;</span>)</a></code></pre></div>
<pre><code>##  [1] add1           alias          anova          case.names    
##  [5] coerce         confint        cooks.distance deviance      
##  [9] dfbeta         dfbetas        drop1          dummy.coef    
## [13] effects        extractAIC     family         formula       
## [17] hatvalues      influence      initialize     kappa         
## [21] labels         logLik         model.frame    model.matrix  
## [25] nobs           plot           predict        print         
## [29] proj           qr             residuals      rstandard     
## [33] rstudent       show           simulate       slotsFromS3   
## [37] summary        variable.names vcov          
## see &#39;?methods&#39; for accessing help and source code</code></pre>
</div>
<p>And lets use some of functions available for class “lm” to extract useful output.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb40-1" data-line-number="1"><span class="kw">coef</span>(lm1)</a>
<a class="sourceLine" id="cb40-2" data-line-number="2"><span class="kw">summary</span>(lm1)</a>
<a class="sourceLine" id="cb40-3" data-line-number="3"><span class="kw">summary</span>(lm1)<span class="op">$</span>coefficients[,<span class="dv">4</span>]</a></code></pre></div>
<pre><code>##     (Intercept) gen_counts[1, ] 
##      46.0159833       0.1007516 
## 
## Call:
## lm(formula = y ~ gen_counts[1, ])
## 
## Residuals:
##     Min      1Q  Median      3Q     Max 
## -31.479 -17.016  -7.212  22.575  34.984 
## 
## Coefficients:
##                 Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept)      46.0160     6.9952   6.578 1.23e-05 ***
## gen_counts[1, ]   0.1008     0.1844   0.546    0.593    
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 23.63 on 14 degrees of freedom
## Multiple R-squared:  0.02088,    Adjusted R-squared:  -0.04906 
## F-statistic: 0.2985 on 1 and 14 DF,  p-value: 0.5934
## 
##     (Intercept) gen_counts[1, ] 
##    1.232512e-05    5.934069e-01</code></pre>
</div>
</div>
<div id="many-linear-regression" class="section level2">
<h2><span class="header-section-number">5.4</span> Many Linear regression</h2>
<p>Just like we did for t-test, we would like to run linear regression for all our genes and extract output in a nice format.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb42-1" data-line-number="1"><span class="kw">apply</span>(gen_counts,<span class="dv">1</span>,<span class="cf">function</span>(x) <span class="kw">lm</span>(y<span class="op">~</span>x))</a></code></pre></div>
</div>
<p>Store this output in a list and have a look at the summary of results fro the first gene.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb43-1" data-line-number="1">lms &lt;-<span class="st"> </span><span class="kw">apply</span>(gen_counts,<span class="dv">1</span>,<span class="cf">function</span>(x) <span class="kw">lm</span>(y<span class="op">~</span>x))</a>
<a class="sourceLine" id="cb43-2" data-line-number="2"><span class="kw">summary</span>(lms[[<span class="dv">1</span>]])</a></code></pre></div>
<pre><code>## 
## Call:
## lm(formula = y ~ x)
## 
## Residuals:
##     Min      1Q  Median      3Q     Max 
## -31.479 -17.016  -7.212  22.575  34.984 
## 
## Coefficients:
##             Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept)  46.0160     6.9952   6.578 1.23e-05 ***
## x             0.1008     0.1844   0.546    0.593    
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 23.63 on 14 degrees of freedom
## Multiple R-squared:  0.02088,    Adjusted R-squared:  -0.04906 
## F-statistic: 0.2985 on 1 and 14 DF,  p-value: 0.5934</code></pre>
</div>
<p>Once you know the output well enough you can try to run <code>lm</code> on all genes and extract output.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb45-1" data-line-number="1">coefs &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">apply</span>(gen_counts,<span class="dv">1</span>,<span class="cf">function</span>(x) <span class="kw">lm</span>(y<span class="op">~</span>x)<span class="op">$</span>coef))</a>
<a class="sourceLine" id="cb45-2" data-line-number="2"></a>
<a class="sourceLine" id="cb45-3" data-line-number="3">li_mo &lt;-<span class="st"> </span><span class="kw">apply</span>(gen_counts,<span class="dv">1</span>,<span class="cf">function</span>(x) <span class="kw">lm</span>(y<span class="op">~</span>x))</a>
<a class="sourceLine" id="cb45-4" data-line-number="4"></a>
<a class="sourceLine" id="cb45-5" data-line-number="5">li_mo_sum &lt;-<span class="st"> </span><span class="kw">apply</span>(gen_counts,<span class="dv">1</span>,<span class="cf">function</span>(x) <span class="kw">summary</span>(<span class="kw">lm</span>(y<span class="op">~</span>x))<span class="op">$</span>coefficients[,<span class="dv">4</span>])</a></code></pre></div>
</div>
</div>
<div id="logistic-regression" class="section level2">
<h2><span class="header-section-number">5.5</span> Logistic regression</h2>
<p>If the outcome variable is categorical, one can use <code>glm()</code> function. We can learn more about this in the future sessions. Just as a teaser, try following.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb46-1" data-line-number="1">y1 &lt;-<span class="st"> </span><span class="kw">factor</span>(pheno<span class="op">$</span>Sex)</a>
<a class="sourceLine" id="cb46-2" data-line-number="2">mylogit &lt;-<span class="st"> </span><span class="kw">glm</span>(y1 <span class="op">~</span><span class="st"> </span>y , <span class="dt">family =</span> <span class="st">&quot;binomial&quot;</span>)</a></code></pre></div>
</div>
</div>
<div id="multiple-testing" class="section level2">
<h2><span class="header-section-number">5.6</span> Multiple testing</h2>
<p>Be aware of multiple testing. We will cover that in the future sessions.</p>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
